PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+		hMenu = _hFolderMenu;
+	else if (nodeType == nodeType_baseDir)
+		hMenu = _hDirectoryMenu;
+	else if (nodeType == nodeType_dir || nodeType == nodeType_dirFile)
+		hMenu = NULL; // currently no context menu for monitored files
+	else //nodeType_file
+		hMenu = _hFileMenu;
+	return hMenu;
+
+}
+
+void ProjectPanel::onTreeItemAdded(bool afterClone, HTREEITEM hItem, TreeViewData* newData)
+{
+	afterClone;
+
+	ProjectPanelData* tvInfo = static_cast<ProjectPanelData*> (newData);
+	tvInfo->setItem( hItem );
+	if (tvInfo->isBaseDirectory())
+		tvInfo->watchDir(true);
+}
+
+void ProjectPanel::onMessage(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+	wParam;
+
+	switch (message)
+	{
+		case DIRECTORYWATCHER_UPDATE:
+		{
+			HTREEITEM hItem = (HTREEITEM)lParam;
+			if (_treeView.itemValid(hItem))
+			{
+
+				TVITEM tvItem;
+				tvItem.hItem = hItem;
+				tvItem.mask = TVIF_PARAM;
+				SendMessage(hwnd, TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+				treeItemChanged(hItem, (TreeViewData*) tvItem.lParam);
+			}
+		}
+		break;
+	}
+
+}
+
+void ProjectPanel::treeItemChanged(HTREEITEM hTreeItem, TreeViewData* tvdata)
+{
+
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hTreeItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+
+	ProjectPanelData& data = *static_cast<ProjectPanelData*>(tvdata);
+	int iImg;
+	switch (data._nodeType)
+	{
+		case nodeType_root:
+			iImg = _isDirty ? INDEX_DIRTY_ROOT : INDEX_CLEAN_ROOT;
+			break;
+		case nodeType_project:
+			iImg = INDEX_PROJECT;
+			break;
+		case nodeType_folder:
+			iImg = (tvItem.state & TVIS_EXPANDED) ? INDEX_OPEN_NODE : INDEX_CLOSED_NODE;
+			break;
+		case nodeType_file:
+			iImg = ::PathFileExists(data._filePath.c_str()) ? INDEX_LEAF : INDEX_LEAF_INVALID;
+			break;
+		case nodeType_dirFile:
+			iImg = INDEX_LEAF_DIRFILE;
+			break;
+		case nodeType_baseDir:
+		case nodeType_dir:
+			rebuildDirectoryTree(hTreeItem, data);
+			return;
+		default:
+			return;
+	}
+	_treeView.setItemImage(hTreeItem, iImg, iImg);
+
+}
+
+generic_string ProjectPanel::buildDirectoryName(const generic_string& filePath, const std::vector<generic_string>& filters)
+{
+	generic_string result(filePath);
+
+	const size_t lastSlashIdx = result.find_last_of(TEXT("\\/"));
+	if (std::string::npos != lastSlashIdx)
+	{
+		result.erase(0, lastSlashIdx + 1);
+		if( result.empty() ) // drive
+		{
+			result = filePath;
+			result.erase(lastSlashIdx);
+		}
+	}
+
+	if (!filters.empty())
+	{
+		result += TEXT(" (");
+		result += stringJoin(filters, TEXT(";"), true);
+		result += TEXT(")");
+	}
+
+	return result;
+
+}
+
+void ProjectPanel::itemVisibilityChanges(HTREEITEM hItem, bool visible)
+{
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.